components/feature/downloads/src/main/java/mozilla/components/feature/downloads/manager/FetchDownloadManager.kt

```diff
@@ -81,29 +84,31 @@ class FetchDownloadManager<T : AbstractFetchDownloadService>(
      * Remove all the listeners.
      */
     override fun unregisterListeners() {
-        if (isSubscribedReceiver) {
-            broadcastManager.unregisterReceiver(this)
-            isSubscribedReceiver = false
+        if (coroutineScope != null) {
+            coroutineScope?.cancel()
+            coroutineScope = null
             queuedDownloads.clear()
         }
     }
 
-    private fun registerBroadcastReceiver() {
-        if (!isSubscribedReceiver) {
-            val filter = IntentFilter(ACTION_DOWNLOAD_COMPLETE)
-            broadcastManager.registerReceiver(this, filter)
-            isSubscribedReceiver = true
+    private fun registerStoreObserver() {
+        coroutineScope?.cancel()
+        coroutineScope = browserStore.flowScoped { flow ->
+            flow.map { it.completedDownloads }
+                    .distinctUntilChanged()
+                    .collect { onReceiveCompletedDownloads(it) }
         }
     }
 
     /**
-     * Invoked when a download is complete. Calls [onDownloadCompleted] and unregisters the
-     * broadcast receiver if there are no more queued downloads.
+     * Finds the completed download in the browser state, then removes it
      */
-    override fun onReceive(context: Context, intent: Intent) {
+    private fun onReceiveCompletedDownloads(completedDownloads: Set<Long>) {
         if (queuedDownloads.isEmpty()) unregisterListeners()
 
-        val downloadID = intent.getLongExtra(EXTRA_DOWNLOAD_ID, -1)
+        val downloadID = completedDownloads
+                .find { queuedDownloads.indexOfKey(it) >= 0 } ?: return
+
```

Comment:
We process a download id from the set here but we never remove it from the set. I think we should add a "consume" action to remove a processed id from the store again.